### PR TITLE
Making Detector Configuration a string and adding TestStand(TS) option

### PR DESCRIPTION
**Pull request by Eikenbcr**
- Made geometry/detector_configuration a string option that allows for HD,VD,TS detectors (others could be added easily enough)
- Added the TS dimensions to DetectorConstruction.cpp
- Made bool->String where needed

### CODE REVIEW

**Comment by Eikenbcr** [comment]
I've already tested this with the various options on a single muon sample,

"/geometry/detector_configuration HD" correctly used the HD config
"/geometry/detector_configuration VD" correctly used the VD config
"/geometry/detector_configuration TS" correctly used the TS config

"/geometry/detector_configuration whatever" gave an error and didn't run the simulation:
parameter value (whatever) is not listed in the candidate List.
***** Illegal parameter (0) </geometry/detector_configuration whatever> *****

Not using the parameter correctly defaulted to the HD configuration.

File: src/ConfigManager.h [hunk 1]
```diff
@@ -69,7 +69,7 @@ class ConfigManager {
     static G4double      GetEventWindow()                   { return Instance()->eventWindow_; }
     static G4bool        GetSNTimingOn()                    { return Instance()->snTimingOn_; }
     static G4String      GetTh2Name()                       { return Instance()->th2Name_; }
-    static G4bool        GetUseHDDetectorConfiguration()    { return Instance()->useHDDetectorConfiguration_; }
+    static G4String      GetDetectorConfiguration()         { return Instance()->detectorConfiguration_; }
```
Comment (kelseymh) [inline src/ConfigManager.h]: Always return non built-in data types by const reference, unless you have a *really* good reason to have to return by value.  Use `const G4String&`.

File: src/ConfigManager.h [hunk 2]
```diff
@@ -106,7 +106,7 @@ class ConfigManager {
     static void SetEventWindow(G4double value)              { Instance()->eventWindow_ = value; }
     static void SetSNTimingOn(G4bool value)                 { Instance()->snTimingOn_ = value; }
     static void SetTh2Name(G4String value)                  { Instance()->th2Name_ = value; }            
-    static void SetUseHDDetectorConfiguration(G4bool value) { Instance()->useHDDetectorConfiguration_ = value; }
+    static void SetDetectorConfiguration(G4String value)    { Instance()->detectorConfiguration_ = value; }
```
Comment (kelseymh) [inline src/ConfigManager.h]: For non-built-in data types, pass by reference, never by value (unless you have a really good reason to need by-value passing).  Here, use `const G4String& value`.

File: src/DetectorConstruction.cpp [hunk 3]
```diff
@@ -63,7 +73,15 @@ G4VPhysicalVolume* DetectorConstruction::Construct()
...
-  std::cout << " Detector configuration is: " << ConfigManager::GetUseHDDetectorConfiguration() << std::endl;
+  if (ConfigManager::GetDetectorConfiguration() == "HD") {
+      std::cout << " Detector configuration is: HD" << std::endl;
+  } else if (ConfigManager::GetDetectorConfiguration() == "VD") {
+      std::cout << " Detector configuration is: VD" << std::endl;
+  } else if (ConfigManager::GetDetectorConfiguration() == "TS") {
+      std::cout << " Detector configuration is: TS" << std::endl;
+  } else {
+      std::cout << " Detector configuration is: DEFAULT (HD)" << std::endl;
+  }
```
Comment (kelseymh) [inline src/DetectorConstruction.cpp]: Why not just one line?
```
std::cout << " Detector configuration is: " << ConfigManager::GetDetectorConfiguration() << std::endl;
```

**Review by kelseymh — CHANGES_REQUESTED** [review]
Looks good!   If it was me, I'd still replace the explicit code in the if-cascade (and the instantiation of solid and LV) with member functions for each detector type.  That would allow for more complicated kinds of detector structures in the future.

File: src/DetectorConstruction.cpp (hunk 3 above)
Comment (Eikenbcr) [reply src/DetectorConstruction.cpp]: That would certainly be easier. Will change this.

File: src/ConfigManager.h (hunk 1 above)
Comment (Eikenbcr) [reply src/ConfigManager.h]: Will revise this line. Thank you Mike!

File: src/ConfigManager.h (hunk 2 above)
Comment (Eikenbcr) [reply src/ConfigManager.h]: Will revise this line. Thank you Mike!

**Comment by Eikenbcr** [comment]
@kelseymh I fixed the issues you pointed out with the code. I also updated the detector geometries to void functions. I had some problems making the world and detector materials and positions as member functions. I wasn't able to figure out how to make their outputs global so I can use them in G4VPhysicalVolume* DetectorConstruction::Construct().

**Comment by kelseymh** [comment]
> @kelseymh I fixed the issues you pointed out with the code. I also updated the detector geometries to void functions.

Oh, I was thinking of something slightly different; maybe we should talk in Skype and I can show you some of the CDMS code.  If you put each detector type into a "`G4LogicalVolume* BuildHDDetector() const {...}`" function, then the main `Construct()` can call that function, stash the return value into a local variable, and then in the main `Construct()` do the placement into the world.

> I had some problems making the world and detector materials and positions as member functions.

The detector material shouldn't necessarily need to be a member "function."  At best, you make a _data member_ of the class, like `G4String detectorMaterial`, and then the individual BuildXYZDetector() functions can reference it to instantiate the right `G4Material*`.  You could even (if you're overly ambitious!) put the active material into ConfigManager and let users change from argon to something else if they wanted to.  That's probably asking for trouble, though.

There's no reason that the detector builder functions need access to the world parameters.  That's the responsibility of `Construct()`.

**Comment by Eikenbcr** [comment]
Okay, lets meet tomorrow to talk about it more. I've reverted back to what I originally had with the fixed print statement because I don't fully trust/understand my void function implementation.

**Comment by Eikenbcr** [comment]
I've pulled some changes that change detector geometries into functions.

**Pull request merged**